### ISSUE TITLE
Slow gradients through growth ODE

### ISSUE THREAD

**Issue by modichirag** [DifferentiableUniverseInitiative/flowpm#82]
I am finding that estimating gradients through the growth ODE are slow, taking almost ~15 seconds, which is about the same time as estimating gradients of 5 step PM simulation on a 128^3 grid. 
This can prove to be a quite cumbersome. 

I have attached the code that estimates and compares time for various gradient estimations, including a case without growth ODE to highlight that this is the bottleneck. 

If the tests are right, then some thoughts -
- we have not implemented growth ODE efficiently, maybe it is reconstructing the graph for that step again and again
(though this seens unlikely since we see that first iteration which generates graph does take much longer...maybe only the ODE part of the graph is getting regenerated again and again)
- maybe TF ODE solver is just this slow 
- one thing to note is that we solve the ODE on 256 'a' (time) points but need the value on only the 5 time-steps we do the simulation. Maybe some workaround can be constructed here. 
- Crude, but we can generate and save an interpolating function for growth and other relevant ODEs. In lambda-CDM, each of these is in  2 dimensional space, depending only on omega_m and h.

Note:
[time_grads2.py.txt](https://github.com/DifferentiableUniverseInitiative/flowpm/files/6541010/time_grads2.py.txt)
Not sure how to add code here..so adding it as txt file. Also, to run it, check the paths for imports, especially the power spectrum code needs to be imported since afaik, it's not part of flowpm yet.

**Comment by modichirag** [comment]
Some timing outputs from the code on 32^3 run

time for grads of Final PS wrt IC in run 0 :  18.782506227493286
time for grads of Final PS wrt IC in run 1 :  8.170791625976562
time for grads of Final PS wrt IC in run 2 :  8.081626176834106

time for grads of Final PS wrt cosmo params in run 0 :  61.804285764694214
time for grads of Final PS wrt cosmo params in run 1 :  25.85694169998169
time for grads of Final PS wrt cosmo params in run 2 :  25.236640214920044

time for grads of pklin wrt cosmo params in run 0 :  24.616602420806885
time for grads of pklin wrt cosmo params in run 1 :  15.966662883758545
time for grads of pklin wrt cosmo params in run 2 :  14.899243831634521

time for grads of pklin without growth function hack, wrt cosmo params in run 0 :  1.5209178924560547
time for grads of pklin without growth function hack, wrt cosmo params in run 1 :  0.008821725845336914
time for grads of pklin without growth function hack, wrt cosmo params in run 2 :  0.0066986083984375

**Comment by modichirag** [comment]
I tried Runge Kutta 4 implementation and it seems to improve things, though graph generation seems to be slow so first iteration takes forever. 

Runge-Kutta 4  implementation is with a simple for loop, I doubt it is very optimized. 
```
@tf.function
def rk4(dydt, a, y0, **kwcosmo):
    da = a[1:] - a[:-1]
    y = []
    y.append(y0)
    steps = a.shape[0]
    for i in range ( 0, steps-1):
        f1 = dydt ( a[i], y[i], **kwcosmo )
        f2 = dydt ( a[i] + da[i] / 2.0, y[i] + da[i] * f1 / 2.0 , **kwcosmo)
        f3 = dydt ( a[i] + da[i] / 2.0, y[i] + da[i] * f2 / 2.0 , **kwcosmo)
        f4 = dydt ( a[i] + da[i],       y[i] + da[i] * f3 , **kwcosmo)
        f =  y[i] + da[i] * ( f1 + 2.0 * f2 + 2.0 * f3 + f4 ) / 6.0
        y.append(f)
    y =  tf.concat(y, axis=0)
    return y
```

Modified code for growth ODE to return tensors instead of a list to be compatible with the rk4 implementaiton
```
# Equation 1.96 from Florent Leclercq thesis                                                                                                                  
@tf.function
def growth_ode(a, y, **kwcosmo):
  r"""Define the ode functions that will be used to compute the linear                                                                                        
  growth factor D_1(a) and second-order growth factor D_2(a) at a given                                                                                       
  scale factor.                                                                                                                                                                                                                                                                                                             
  """
  a = tf.convert_to_tensor(a, dtype=tf.float32)
  # Instantiate a cosmology object                                                                                                                            
  cosmo = Cosmology(**kwcosmo)

  # Extracting entries                                                                                                                                        
  d1, d1_f, d2, d2_f = tf.split(y, 4, axis=1)

  # ODE for d1                                                                                                                                                
  dy1dt = d1_f, 1.5 * Omega_m_a(cosmo, a) * d1 / tf.pow(a, 2) - (d1_f / a) * (
    Omega_de_a(cosmo, a) - 0.5 * Omega_m_a(cosmo, a) + 2)
  # ODE for d2                                                                                                                                                
  dy2dt = d2_f, 1.5 * Omega_m_a(cosmo, a) * d2 / tf.pow(a, 2) - (
      d2_f / a) * (Omega_de_a(cosmo, a) - 0.5 * Omega_m_a(cosmo, a) +
                   2) - 1.5 * (Omega_m_a(cosmo, a) * d1**2) / tf.pow(a, 2)

  # Concatenate output                                                                                                                                        
  dydt = [dy1dt[0], dy1dt[1], dy2dt[0], dy2dt[1]]
  return tf.concat(dydt, axis=1)
```

**Comment by modichirag** [comment]
Timings for 256 steps
```
For DP solver
Time for grads of pklin wrt cosmo params in run 0 :  19.48847723007202
time for grads of pklin wrt cosmo params in run 1 :  9.085368633270264
time for grads of pklin wrt cosmo params in run 2 :  9.25782322883606
For RK4
time for grads of pklin wrt cosmo params in run 0 :  754.3619964122772
time for grads of pklin wrt cosmo params in run 1 :  0.5503551959991455
time for grads of pklin wrt cosmo params in run 2 :  0.5490517616271973
```

However we should be able to get away with 32 or 64 steps as the changes in the function are <0.5% I think
For 64 steps
```
For Durmond Prince
time for grads of pklin wrt cosmo params in run 0 :  12.9021155834198
time for grads of pklin wrt cosmo params in run 1 :  3.299262523651123
time for grads of pklin wrt cosmo params in run 2 :  3.0157277584075928
For RK4
time for grads of pklin wrt cosmo params in run 0 :  141.93882060050964
time for grads of pklin wrt cosmo params in run 1 :  0.14385175704956055
time for grads of pklin wrt cosmo params in run 2 :  0.14376211166381836
```

**Comment by EiffL** [comment]
Thanks for this @modichirag !!! So, one comment, your implementation of rk4 will work but will be verry slow to compile, because it creates a large graph. You want to avoid having to do a python for loop.
This is why I was using a lax.scan in jax. In TensorFlow I think you can use a https://www.tensorflow.org/api_docs/python/tf/while_loop

It will greatly improve the compile time.

**Comment by modichirag** [comment]
Ah, thanks for pointing that out. I had tried while loop for RIM instead of a for loop, and did not find any differences in timing but maybe because the bottleneck was somewhere else. Hence I just went for "for" loop here since it was convenient. 

Here is new while_loop implementation

```
@tf.function
def rk4(dydt, a, y0, **kwcosmo):

    da = a[1:] - a[:-1]
    steps = len(da)
    y = tf.TensorArray(size=steps+1, dtype=tf.float32)
    y = y.write(0, y0)
    def body(i, yt, y):
        t = a[i]
        dt = da[i]
        f1 = dydt ( t,            yt , **kwcosmo )
        f2 = dydt ( t + dt / 2.0, yt + dt * f1 / 2.0 , **kwcosmo )
        f3 = dydt ( t + dt / 2.0, yt + dt * f2 / 2.0 , **kwcosmo )
        f4 = dydt ( t + dt,       yt + dt * f3 , **kwcosmo )
        f =  yt + dt * ( f1 + 2.0 * f2 + 2.0 * f3 + f4 ) / 6.0
        y = y.write(i+1, f)
        return i+1, f, y

    def cond(i, yt, y):
            return tf.less(i, tf.constant(steps))
    i = 0
    i, yl, y =  tf.while_loop(cond, body, [i, y0, y], swap_memory=True)
    return y.stack()
```

Timings for 256 steps
```
time for grads of pklin wrt cosmo params in run 0 :  11.44162368774414
time for grads of pklin wrt cosmo params in run 1 :  1.13869047164917
time for grads of pklin wrt cosmo params in run 2 :  1.1294705867767334
```
Timings for 64 steps
```
time for grads of pklin wrt cosmo params in run 0 :  10.767854928970337
time for grads of pklin wrt cosmo params in run 1 :  0.34038329124450684
time for grads of pklin wrt cosmo params in run 2 :  0.3020448684692383
```

Graph generation is significantly faster, but every derivative is ~twice as slow.
So there seems to be a trade-off.

**Comment by EiffL** [comment]
so where did we land on this? we also want to make sure that this is fully executed on GPU, so probaly `swap_memory=False`

Also @modichirag , note that you can try to XLA compile the code which is supposed to speed things up: https://www.tensorflow.org/xla#enable_xla_for_tensorflow_models

**Comment by EiffL** [comment]
I would be happy to review a PR ;-)

**Comment by modichirag** [comment]
> so where did we land on this? we also want to make sure that this is fully executed on GPU, so probaly `swap_memory=False`
> 
Yeah, I set it to False in my next commit already, but thanks for pointing it out. 
As to where we land, well, I was wondering if you had any thoughts on the trade-off. 
Assuming 64 steps to solve ODE, we save ~130 seconds in graph generation but  pay 0.15 seconds per call. That may or may not add up.

However I presume the while loop certainly saves on memory and is scalable to more time steps (as well as being cleaner).
So do we move ahead with that?

> Also @modichirag , note that you can try to XLA compile the code which is supposed to speed things up: https://www.tensorflow.org/xla#enable_xla_for_tensorflow_models

https://www.tensorflow.org/xla/known_issues
I think that piece of code falls for half of these issues.
While fixing the number of steps for integration beforehand, as we do anyways, solves 2 of them, I do not think we can resolve this- 
_Unimplemented: Support for TensorList crossing the XLA/TF boundary is not implemented_
Can we?
